face_cropping/src/FaceCropper.cpp

```diff
@@ -0,0 +1,468 @@
+
+    if (cutout.height < m_oldCutout.height + minHeightChange && cutout.height > m_oldCutout.height - minHeightChange &&
+        cutout.width < m_oldCutout.width + minWidthChange && cutout.width > m_oldCutout.width - minWidthChange)
+    {
+        cutout.height = m_oldCutout.height;
+        cutout.width = m_oldCutout.width;
+    }
+    if (cutout.y < m_oldCutout.y + minYChange && cutout.y > m_oldCutout.y - minYChange)
+    {
+        cutout.y = m_oldCutout.y;
+    }
+    if (cutout.x < m_oldCutout.x + minXChange && cutout.x > m_oldCutout.x - minXChange)
+    {
+        cutout.x = m_oldCutout.x;
+    }
+    return cutout;
+}
+
+cv::Rect FaceCropper::validateAspectRatio(cv::Rect cutout, cv::Rect face)
+{
+    if (face.height / static_cast<float>(face.width) >= m_aspectRatio)
+    {
+        cutout.height = getClosestNumberDividableBy(cutout.height, m_yAspect / m_xAspect);
+        cutout.width = cutout.height / m_aspectRatio;
+    }
+    else
+    {
+        cutout.width = getClosestNumberDividableBy(cutout.width, m_xAspect / m_yAspect);
+        cutout.height = cutout.width * m_aspectRatio;
+    }
+
+    if (cutout.width < m_oldCutout.width + m_xAspect && cutout.width > m_oldCutout.width - m_xAspect ||
+        cutout.height < m_oldCutout.height + m_yAspect && cutout.height > m_oldCutout.height - m_yAspect)
+    {
+        cutout.width = m_oldCutout.width;
+        cutout.height = m_oldCutout.height;
+    }
+    return cutout;
+}
+
+cv::Rect FaceCropper::correctCutoutBoundaries(cv::Rect cutout, cv::Mat frame)
+{
+    if (cutout.x > frame.size().width)
+    {
+        cutout.x = frame.size().width - cutout.width;
+    }
+    else if (cutout.x < 0)
+    {
+        cutout.x = 0;
+    }
+
+    if (cutout.width <= 0)
+    {
+        cutout.width = 1;
+    }
+    else if (cutout.width + cutout.x > frame.size().width)
+    {
+        cutout.x = frame.size().width - cutout.width;
+    }
+
+    if (cutout.y > frame.size().height)
+    {
+        cutout.y = frame.size().height - cutout.height;
+    }
+    else if (cutout.y < 0)
+    {
+        cutout.y = 0;
+    }
+
+    if (cutout.height <= 0)
+    {
+        cutout.height = 1;
+    }
+    else if (cutout.height + cutout.y > frame.size().height)
+    {
+        cutout.y = frame.size().height - cutout.height;
+    }
+    return cutout;
+}
+
+cv::Rect FaceCropper::getAverageRect(std::list<cv::Rect> list)
+{
+    cv::Rect r(0, 0, 0, 0);
+    if (!list.empty())
+    {
+        int i = 0;
+        for (cv::Rect& rect : list)
+        {
+            i++;
+            r.x += rect.x;
+            r.y += rect.y;
+            r.width += rect.width;
+            r.height += rect.height;
+        }
+        r.x /= i;
+        r.y /= i;
+
+        float newWidth = r.width / i;
+        if (static_cast<int>(trunc(newWidth)) > 0)
+        {
+            r.width = newWidth;
+        }
+        else
+        {
+            r.width = static_cast<int>(round(newWidth));
+        }
```

Comment:
I am not sure why you are using round here. I understand that you don't want the width to be 0. 
Is the following code equivalent?
```suggestion
        if (newWidth >= 1.f)
        {
            r.width = newWidth;
        }
        else
        {
            r.width = 1;
        }
```

Reply:
Yes, it would and is the `static_cast` necessary, would an implicit conversion do the same thing?

Reply:
Without static_cast, you would get a warning.